source/connections-survive-step-down/tests/README.rst

```diff
@@ -0,0 +1,214 @@
+   matches the address of the server E, set ``priority`` to 1. Otherwise
+   set ``priority`` to 0.
+3. Increment the ``version`` field of the replica set configuration.
+4. Reconfigure the replica set using the `replSetReconfig
+   <https://docs.mongodb.com/manual/reference/command/replSetReconfig/>`_
+   command, providing the modified configuration.
+5. Directly connect to the server N and execute `replSetFreeze
+   <https://docs.mongodb.com/manual/reference/command/replSetFreeze/>`_
+   admin command as follows: ``{replSetFreeze: 0}``. If this command
+   fails with operation failure code 95 ("cannot freeze node when primary or
+   running for election. state: Primary"), perform server selection to
+   discover the current primary. If the current primary is server N, stop
+   as the procedure is complete. Otherwise propagate the operation failure
+   error(*).
+6. Connect to the server E and execute the `replSetStepDown
+   <https://docs.mongodb.com/manual/reference/command/replSetStepDown/>`_
+   admin command as follows:
+   ``{replSetStepDown: 4, secondaryCatchUpPeriodSecs: 2}``.
+7. Connect directly to the server N and execute the following admin command:
+   ``{replSetStepUp: 1}``. If this command fails with an operation failure
+   error with code 125 ("Election failed"), repeat this step 7.
+8. Set the status of all servers in the topology to Unknown.
+9. Perform server selection to obtain the current primary.
+10. If the current primary is not N, go to step 7 and step up the server N
+    again.
+
+(*) The cluster may have held an election after the replica set was reconfigured
+in step 4, making the server N the current primary. If this happens,
+``replSetFreeze`` command will fail, but the overall goal of having the
+server N as the primary has been achieved.
+
+As the above procedure contains a potentially infinite loop, the driver MAY
+impose a time limit and fail if the server N has not become a primary in the
+allotted time. This time limit SHOULD be at least 10 seconds.
+
+Stepping Down A Server
+``````````````````````
+
+If the driver does not care which server becomes the new primary, as would be
+the case in the getMore test for example, the driver should perform the
+following procedure:
+
+- Obtain a list of servers in the cluster.
+- Randomly or otherwise choose a server to be the new primary, other than the
+  current primary.
+- Follow the above server election procedure to elect the chosen server as
+  the new primary.
+
+Tests
+-----
+
+The driver should implement the following tests:
+
+getMore Iteration
+`````````````````
+
+This test requires server version 4.2 or higher.
+
+Perform the following operations:
+
+- Insert 5 documents into a collection.
+- Start a find operation on the collection with a batch size of 2, and
+  retrieve the first batch of results.
+- Step down the primary, and verify that another server has been elected as
+  a primary.
```

Comment:
Is it actually necessary to assert that another server has been elected? If we issue a find query on a primary and then step it down, I would think we could just continue iterating to ensure that the driver successfully issues a getMore against the former primary. Provided the 4.2 server isn't dropping its connections on a step down, I don't see what this test has to do with SDAM.

Reply:
Consider the following scenario:

- Server A is the primary. We step it down.
- Server B is currently not a primary. The driver notices this and sends a getMore to the cluster.
- By the time the cluster receives the getMore, server A has gotten elected a primary again (instead of some other server).
- The getMore ended up getting routed to the current primary.

One solution to this would be to check server state after getMore and verify it is still not a primary, if you find this easier than ensuring that some other server is a primary before sending the getMore.

Reply:
I wouldn't think a test would need to worry about A being reëlected after a step down, as [`replSetStepDown`](https://docs.mongodb.com/manual/reference/command/replSetStepDown/) applies a 60-second waiting period by default. That said, I don't see the harm in ensuring the SDAM picks up a new primary before we continue iteration on the cursor. IIRC, cursors remain pinned to the original server so we wouldn't have to worry about a getMore inadvertently targeting a different node. Also, retryable reads also don't come into play for getMore operations, so there's nothing to worry about there.

Reply:
The procedure described reduces that 60 second period to 4 seconds, and thus should not be relied on.

Reply:
Ah, I just saw the `4` above. Is that value intended to allow the RS to restore itself to the original state before the subsequent test? I don't recall reading anything in the document about ensuring the RS is in an original state before running each test (perhaps I missed it), but even 4 seconds seems like it might be too long in some cases.

This reminds me of similar challenges in Emily's original [failover integration tests](https://github.com/estolfo/mo-tests) (and relevant to our upcoming failover testing spec). I believe the solution there was to use a freshly constructed replica set for each test, so there was no concern about the leftover state from the previous test.

Do you think that would be preferable here, or would it be too much work to expect drivers to coordinate with MO during the test suite?